docs/typescript.md

```diff
@@ -2,13 +2,504 @@
+
+If that is not the case, however, mocking only the properties needed in the test improves readability. It makes the intention and scope of the mock clear, and is much more convenient to write.
+
+### Compiler Directives
+
+<!-- TODO: Add section for `@ts-expect-error` -->
+
+#### Avoid `any`
+
+`any` is the most dangerous form of explicit type declaration, and should be completely avoided if possible.
+
+- `any` doesn't represent the widest type, or indeed any type at all. `any` is a compiler directive for _disabling_ static type checking for the value or type to which it's assigned.
+- `any` suppresses all error messages about its assignee. This includes errors that are changed or newly introduced by alterations to the code. This makes `any` the cause of dangerous **silent failures**, where the code fails at runtime but the compiler does not provide any prior warning.
+- `any` subsumes all other types it comes into contact with. Any type that is in a union, intersection, is a property of, or has any other relationship with an `any` type or value is erased and becomes an `any` type itself.
+
+```typescript
+// Type of 'payload_0': 'any'
+const handler:
+  | ((payload_0: ComposableControllerState, payload_1: Patch[]) => void)
+  | ((payload_0: any, payload_1: Patch[]) => void);
+```
+<!-- TODO: Add more examples: For instance, if you have a function that is declared to return any which actually returns an object, all properties of that object will be any. -->
+
+- Because of this, `any` infects all surrounding and downstream code with its directive to suppress errors. This is the most dangerous characteristic of `any`, as it expands the surface area of unchecked code for which the TypeScript compiler can make no guarantees regarding type safety or runtime behavior.
+
+##### Try `unknown` and `never` instead
+
+###### `unknown`
+
+- `unknown` is the universal supertype i.e. the widest possible type, equivalent to the universal set(U).
+- Every type is assignable to `unknown`, but `unknown` is only assignable to `unknown`.
+- When typing the _assignee_, `any` and `unknown` are completely interchangeable since every type is assignable to both.
+- `any` usage is often motivated by a need to find a placeholder type that could be anything. `unknown` is the most likely type-safe substitute for `any` in these cases.
+
+###### `never`
+
+- `never` is the universal subtype i.e. the narrowest possible type, equivalent to the null set(∅).
+- `never` is assignable to every type, but the only type that is assignable to `never` is `never`.
+- When typing the _assigned_:
+  - `unknown` is unable to replace `any`, as `unknown` is only assignable to `unknown`.
+  - The type of the _assigned_ must be a subtype of the _assignee_.
+  - `never` is worth trying, as it is the universal subtype and assignable to all types.
+
+<!-- TODO: Add examples -->
+
+##### Don't allow `any` to be used as a generic default argument
```

Comment:
Q: wouldn't it be enough to just keep the paragraph explaining any must not be used because of its disabling of type checks and not add more? There's too many text about do not use any for this or than when just do not use any first explanation is probably enough.

Reply:
We do have a eslint rule for banning `any` usage now, and [usually](https://github.com/MetaMask/contributor-docs/pull/80#discussion_r1507074598) I would agree that conventions that we enforce with linters or static code analysis do not require detailed explanations or justifications.

However, I strongly feel that `any` is a special case that warrants special attention.
- The epics and issues attached to this PR should show that writing guidelines for `any` usage is the main scope of this ticket.
- The tickets attached to those epics will show that a significant portion of the work shared libraries does is fixing existing `any` usage and preventing introduction of new instances through code review.
- This is all because `any` is such a convenient escape hatch that there is a strong incentive to override the eslint ban and use it anyway. Simply reiterating the ban is not enough to be effective.
- Therefore, the guidelines must be written to cultivate a deeper understanding in our contributors about exactly why and how `any` usage is destructive to a codebase.

Reply:
I added a short explanation summarizing the above: https://github.com/MetaMask/contributor-docs/pull/74/files#diff-ae03a9f2eb638a25c1d8d7ba853c78220f1ed44a19ad97a8d780cbc0c00a4260R411-R413